### ISSUE TITLE
Python Test Failures 

### ISSUE THREAD

**Issue by Creling** [lightbulb128/troy-nova#10]
Hi,

I encountered a problem where nearly all Python tests fail in my environment, including some basic operations.

![image](https://github.com/user-attachments/assets/fa4697e4-b482-4a5d-87e1-8b968370949f)

I suspect the issue might be related to the decode function.

```python
# test_he_operations.py

def test_add_sub(self):
    ghe = self.ghe
    message1 = ghe.random_simd_full()
    message2 = ghe.random_simd_full()
    plain1 = ghe.encoder.encode_simd(message1)
    plain2 = ghe.encoder.encode_simd(message2)
    cipher1 = ghe.encryptor.encrypt_symmetric_new(plain1, False)
    cipher2 = ghe.encryptor.encrypt_symmetric_new(plain2, False)
    added = ghe.evaluator.add_new(cipher1, cipher2)
    decoded = ghe.encoder.decode_simd(ghe.decryptor.decrypt_new(added))

    print(f"message1: {message1}")
    print(f"message2: {message2}")
    print(f"decoded: {decoded}")

'''
outputs:

message1: [ 105270  500069  262731  803957  543788  756142  244039 1036853   65449
  229446  656679   88176  384052  776630  392070  912476  543635  258704
  999613  962219  939314  428918  597307  417538  478815 1037033   92833
  119667  836193  217219  852762   98650]
message2: [ 199749  421437  145283  560030  414781  466479  338126  535338   96947
  902743 1028024  940739  261747  622940  455635  888047  416033  356917
  242463  823168  747736  339538  860854  887122  584590  205416  329272
  430375  342742   29770    5727  702475]
decoded: [305019 305019 305019 305019 305019 305019 305019 305019 305019 305019
 305019 305019 305019 305019 305019 305019 305019 305019 305019 305019
 305019 305019 305019 305019 305019 305019 305019 305019 305019 305019
 305019 305019]
'''
```

My environment details are as follows:

```
Python 3.11.9
CMake  3.30.2
Nvidia Driver 545.23.08    
CUDA 12.3
RTX 3090
```

P.S. I have added the cmake flags mentioned in https://github.com/lightbulb128/troy-nova/issues/1, the C++ tests work well.

![image](https://github.com/user-attachments/assets/46870ae8-9af3-470e-b721-437dc2332856)

**Comment by Creling** [comment]
Well, all decode related tests fails, regardless of whether they are run on the host or device.

```
test_extract_lwe (__main__.HostBFVHeOperations.test_extract_lwe) ... ok
test_setup_ok (__main__.HostBFVHeOperations.test_setup_ok) ... ok
test_extract_lwe (__main__.HostCKKSHeOperations.test_extract_lwe) ... ok
test_setup_ok (__main__.HostCKKSHeOperations.test_setup_ok) ... ok
test_extract_lwe (__main__.HostBGVHeOperations.test_extract_lwe) ... ok
test_setup_ok (__main__.HostBGVHeOperations.test_setup_ok) ... ok
test_extract_lwe (__main__.DeviceBFVHeOperations.test_extract_lwe) ... ok
test_setup_ok (__main__.DeviceBFVHeOperations.test_setup_ok) ... ok
test_extract_lwe (__main__.DeviceCKKSHeOperations.test_extract_lwe) ... ok
test_setup_ok (__main__.DeviceCKKSHeOperations.test_setup_ok) ... ok
test_extract_lwe (__main__.DeviceBGVHeOperations.test_extract_lwe) ... ok
test_setup_ok (__main__.DeviceBGVHeOperations.test_setup_ok) ... ok
```

**Comment by Creling** [comment]
"SerializeTest.DeviceCKKSCiphertext" fails in line 140:

https://github.com/lightbulb128/troy-nova/blob/c89a8980c2b266d9fe82f69b03f52b62abaecf5f/test/serialize.cu#L134-L141

```
Value of: truth.near_equal(decoded, tolerance)
  Actual: false
Expected: true
```

**Comment by lightbulb128** [comment]
"SerializeTest.DeviceCKKSCiphertext" seems unrelated to python's failing where I just need to tune up the tolerance.

Don't know why exactly the python decoding are failing. From the output, I see the decoded vector has all the same value, and the value equals the addition of the first elements in the input vectors. Probably I should check this with exactly the same environment as yours.

**Comment by Creling** [comment]
Yes, "SerializeTest.DeviceCKKSCiphertext" is unrelead to this issue.

For the decoding problem, what is your environment?

**Comment by lightbulb128** [comment]
My environment:
* NVIDIA GeForce RTX 4090
* CUDA 12.4, drivers 550.54.14
* g++/gcc 11.4
* Ubuntu 20.04
* Python 3.8.10
* cmake 3.28.1

**Comment by Creling** [comment]
> My environment:
> 
> * NVIDIA GeForce RTX 4090
> * CUDA 12.4, drivers 550.54.14
> * g++/gcc 11.4
> * Ubuntu 20.04
> * Python 3.8.10
> * cmake 3.28.1

I test in Python 3.8 and it works, all tests are ok.

**Comment by Creling** [comment]
Thank you for your help.

**Comment by lightbulb128** [comment]
Reproduced this issue with python3.10.15. Solve: seems pybind11 version is too old. Updating from pybind11 v2.11 to v2.13 solved the reproduction of the issue (#14).
